### ISSUE TITLE
Lint unconditional recursion in trait implementations

### ISSUE THREAD

**Issue by clarfonthey** [rust-lang/rust-clippy#1825]
This would basically extend the current `unconditional_recursion` lint, except it would be deny by default, and would trigger on any trait implementation that "accidentally" recurses at the definition, like this:

```
trait Stuff {
    fn method(&self) -> usize
}
impl Stuff for Thing {
    fn method(&self) -> usize {
         thing.method()
    }
}
```

Because `Thing` doesn't actually contain a definition for `method`, this is unconditional recursion, hiding a bug.

**Comment by sinkuu** [comment]
This is already caught by the compiler.

```rust
trait Stuff {
    fn method(&self) -> usize;
}

struct Thing;

impl Stuff for Thing {
    fn method(&self) -> usize {
         self.method()
    }
}
```

```
warning: function cannot return without recurring
  --> <anon>:8:5
   |
8  | /     fn method(&self) -> usize {
9  | |          self.method()
10 | |     }
   | |_____^
   |
   = note: #[warn(unconditional_recursion)] on by default
note: recursive call site
  --> <anon>:9:10
   |
9  |          self.method()
   |          ^^^^^^^^^^^^^
   = help: a `loop` may express intention better if this is on purpose
```

**Comment by clarcharr** [comment]
@sinkuu: it is, but the point is that this would be a special case of the compiler test to deny-by-default on certain obvious cases

**Comment by oli-obk** [comment]
I don't think we should be duplicating code from rustc. The compiler lint could be deny by default IMO, but I don't think it really matters, because you shouldn't run code that has warnings anyway...

**Comment by clarfonthey** [comment]
Closing this since rustc handles this case, and it not being deny-by-default isn't the worst.
